pxtblocks/blocklydiff.ts

```diff
@@ -0,0 +1,379 @@
+        }
+
+        function logTodo(msg: string) {
+            log(`${msg}:`, Util.values(todoBlocks).map(b => b.toDevString()))
+        }
+
+        function moved(b: Blockly.Block) {
+            const oldb = oldWs.getBlockById(b.id); // extra block created in added step
+            if (!oldb)
+                return false;
+
+            const newPrevious = b.getPreviousBlock();
+            // connection already already processed
+            if (newPrevious && !todoBlocks[newPrevious.id])
+                return false;
+            const newNext = b.getNextBlock();
+            // already processed
+            if (newNext && !todoBlocks[newNext.id])
+                return false;
+
+            const oldPrevious = oldb.getPreviousBlock();
+            if (!oldPrevious && !newPrevious) return false; // no connection
+            if (!!oldPrevious != !!newPrevious // new connection
+                || oldPrevious.id != newPrevious.id) // new connected blocks
+                return true;
+            const oldNext = oldb.getNextBlock();
+            if (!oldNext && !newNext) return false; // no connection
+            if (!!oldNext != !!newNext // new connection
+                || oldNext.id != newNext.id) // new connected blocks
+                return true;
+            return false;
+        }
+
+        function changed(b: Blockly.Block) {
+            let oldb = oldWs.getBlockById(b.id); // extra block created in added step
+            if (!oldb)
+                return false;
+
+            // normalize
+            //oldb = copyToTrashWs(oldb);
+            const oldText = normalizedDom(oldb);
+
+            //b = copyToTrashWs(b);
+            const newText = normalizedDom(b);
+
+            if (oldText != newText) {
+                log(`old ${oldb.toDevString()}`, oldText)
+                log(`new ${b.toDevString()}`, newText)
+                return true;
+            }
+
+            // not changed!
+            return false;
+        }
+
+        function normalizedDom(b: Blockly.Block, keepChildren?: boolean): string {
+            const dom = Blockly.Xml.blockToDom(b, true);
+            normalizeAttributes(dom);
+            visDom(dom, (e) => {
+                normalizeAttributes(e);
+                if (!keepChildren) {
+                    if (e.localName == "next")
+                        e.remove(); // disconnect or unplug not working propertly
+                    if (e.localName == "statement")
+                        e.remove();
+                }
+            })
+            return Blockly.Xml.domToText(dom);
+        }
+
+        function normalizeAttributes(e: Element) {
+            e.removeAttribute("id");
+            e.removeAttribute("x");
+            e.removeAttribute("y");
+            e.removeAttribute("deletable");
+            e.removeAttribute("editable");
+            e.removeAttribute("movable")
+        }
+
+        function visDom(el: Element, f: (e: Element) => void) {
```

Comment:
is there a better name we could use for this? / could this just be done as `el.queryselectorAll("*").forEach(f)` or something of the sort?